### PR TITLE
[2022-07 LWG Motion 1] P2618R0 C++ Standard Library Issues

### LINKED ISSUE #5588: [2022-07 LWG Motion 1] P2618R0 C++ Standard Library Issues to be moved in Virtual Plenary, Jul. 2022
[P2618R0](https://wg21.link/p2618r0)

### PR DESCRIPTION
**Pull request by burblebee**
Fixes #5588
Fixes cplusplus/papers#1281

### CODE REVIEW

**Comment by burblebee** [comment]
The build is failing in check-output.sh because section depr.atomics.flag was removed by this paper (the failure is in the check for "Cross references since the previous standard.").  How to make the build happy?

**Comment by Mick235711** [comment]
> The build is failing in check-output.sh because section depr.atomics.flag was removed by this paper (the failure is in the check for "Cross references since the previous standard."). How to make the build happy?

I'm guessing add something like
```latex
\removedxref{depr.atomics.flag}
```
into [`xrefdelta.tex`](https://github.com/cplusplus/draft/blob/main/source/xrefdelta.tex) may satisfy the build script.

But again, I'm no expert in draft LaTeX usage, so I'm not sure if this will fix it or not.

File: source/utilities.tex [hunk 1]
```diff
@@ -7977,6 +7971,21 @@
...
+Implementations are not permitted to use additional storage,
+such as dynamic memory, to allocate the object of type \tcode{E}.
+Member \tcode{has_val} indicates whether the \tcode{expected<T, E>} object
+represents a value of type \tcode{T}.
+
+\pnum
+A program that instantiates
+the definition of the template \tcode{expected<T, E>} with
+a type for the \tcode{E} parameter that
+is not a valid template argument for unexpected is ill-formed.
```
Comment (jwakely) [inline source/utilities.tex:7987]:
```suggestion
is not a valid template argument for \tcode{unexpected} is ill-formed.
```

File: source/utilities.tex [hunk 2]
```diff
@@ -7977,6 +7971,21 @@
 };
 \end{codeblock}
 
+\pnum
+Any object of type \tcode{expected<T, E>} either
+represents a value of type \tcode{T}, or
+contains a value of type \tcode{E} within its own storage.
+Implementations are not permitted to use additional storage,
+such as dynamic memory, to allocate the object of type \tcode{E}.
+Member \tcode{has_val} indicates whether the \tcode{expected<T, E>} object
```
Comment (jwakely) [inline source/utilities.tex:7980]:
```suggestion
Member \exposid{has_val} indicates whether the \tcode{expected<T, E>} object
```

File: source/ranges.tex [hunk 3]
```diff
@@ -4514,7 +4521,7 @@
 \tcode{iterator_traits<iterator_t<\exposid{Base}>>::iterator_category}.
 \begin{itemize}
 \item
-If \tcode{is_lvalue_reference_v<invoke_result_t<F\&, range_reference_t<\exposid{Base}>>>}
+If \tcode{is_lvalue_reference_v<invoke_result_t<@\exposidnc{maybe-const}@<Const, F>\&, range_reference_t<\linebreak\exposid{Base}>>>}
```
Comment (jwakely) [inline source/ranges.tex:4524]: These `@` characters appear in the rendered PDF.

```suggestion
If \tcode{is_lvalue_reference_v<invoke_result_t<\exposidnc{maybe-const}<Const, F>\&, range_reference_t<\linebreak\exposid{Base}>>>}
```

File: source/utilities.tex [hunk 4]
```diff
@@ -11676,9 +11685,11 @@
 \pnum
 \constraints
 \tcode{\&F::operator()} is well-formed when treated as an unevaluated operand and
-\tcode{decltype(\brk{}\&F::operator())} is of the form
+\tcode{decltype(\brk{}\&F::operator())} is either of the form
 \tcode{R(G::*)(A...)}~\cv{}~\tcode{\opt{\&}~\opt{noexcept}}
-for a class type \tcode{G}.
+or of the form
+\tcode{R(*)(G, A...)}~\tcode{\opt{noexcept}}
```
Comment (jwakely) [inline source/utilities.tex:11691]: I think closing the `tcode` and reopening it was only needed for the _cv_ macro, which isn't in this new text.

```suggestion
\tcode{R(*)(G, A...) \opt{noexcept}}
```

File: source/threads.tex [hunk 5]
```diff
@@ -11319,9 +11343,11 @@
 \constraints
 \tcode{\&F::operator()} is well-formed when
 treated as an unevaluated operand\iref{term.unevaluated.operand} and
-\tcode{decltype(\brk{}\&F::operator())} is of the form
+\tcode{decltype(\brk{}\&F::operator())} is either of the form
 \tcode{R(G::*)(A...)}~\cv{}~\tcode{\opt{\&}~\opt{noexcept}}
-for a class type \tcode{G}.
+or of the form
+\tcode{R(*)(G, A...)}~\tcode{\opt{noexcept}}
```
Comment (jwakely) [inline source/threads.tex:11349]: I think closing the `tcode` and reopening it was only needed for the _cv_ macro, which isn't in this new text.

```suggestion
\tcode{R(*)(G, A...) \opt{noexcept}}
```

This does not change the rendered PDF.

**Review by jwakely — CHANGES_REQUESTED** [review]
There is one error in the content, a couple of formatting problems, and a couple of no-op changes to simplify the LaTeX.

File: source/utilities.tex (hunk 1 above)
Comment (burblebee) [reply source/utilities.tex:7987]: thanks!

File: source/utilities.tex (hunk 2 above)
Comment (burblebee) [reply source/utilities.tex:7980]: fixed

File: source/ranges.tex (hunk 3 above)
Comment (burblebee) [reply source/ranges.tex:4524]: fixed

File: source/threads.tex (hunk 5 above)
Comment (burblebee) [reply source/threads.tex:11349]: fixed

File: source/utilities.tex (hunk 4 above)
Comment (burblebee) [reply source/utilities.tex]: fixed

**Review by jwakely — APPROVED** [review]
Looks good now, thanks

File: source/threads.tex [hunk 6]
```diff
@@ -11316,9 +11316,11 @@
 \begin{itemize}
 \item
 \tcode{F::operator()} is a non-static member function and
-\tcode{decltype(\brk{}\&F::operator())} is of the form
+\tcode{decltype(\brk{}\&F::operator())} is either of the form
 \tcode{R(G::*)(A...)}~\cv{}~\tcode{\opt{\&}~\opt{noexcept}}
-for a class type \tcode{G}, or
+or of the form
+\tcode{R(*)(G, A...)~\opt{noexcept}}
+for a type \tcode{G}, or
```
Comment (tkoeppe) [inline source/threads.tex:11347]: @brevzin, @jwakely, @CaseyCarter et al: I integrated the incoming wording with the prior change of CWG-12/["static operator()"](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p1169r4.html). Could you please have a look if the merged wording is acceptable?

Comment (tkoeppe) [reply source/threads.tex:11347]: @timsong-cpp also.

Comment (jwakely) [reply source/threads.tex:11347]: LGTM

Comment (jwakely) [reply source/threads.tex:11347]: CWG-12 assumes this was applied, so the end result should be exactly what's shown in CWG-12, and it looks like that is the case.

Comment (tkoeppe) [reply source/threads.tex:11347]: Thank you! Oh yes indeed, there's even an note to the editor in LWG3617 to this effect, perfect!

Comment (tkoeppe) [reply source/threads.tex:11347]: Oh, except that the CWG-17 paper says `G cv-ref_opt` where LWG3617 just says `G`.

Comment (tkoeppe) [reply source/threads.tex:11347]: But I suppose since `G` is just "a type", that can include qualifiers.

**Pull request merged**